### DISCUSSION TITLE
Audio mixing from 2 channels

### DISCUSSION THREAD

**Discussion by MohdIrfan12** [google/oboe#1313] [Q&A] [answered]
Does oboe support audio mixing? An app I am working on has the following use case - 

1. Songs are being played in a music app on a smartphone.
2. App I am working on listens for mic input on the same smartphone. 

I need to combine these 2 audio streams into one and then transmit it onto a channel. 

That the transmitting part has been taken care of what left is just the audio mixing part. I was wondering whether the oboe supports this feature?

Forgive me for labeling it as a bug (it's more of a query), did it by mistake and now I can't find the option to change it.

**Comment by dturner** [comment, marked as answer]
> Does oboe support audio mixing?

The Oboe core codebase does not include any mixing functionality, however, mixing is pretty simple and is demonstrated in some of the samples using the shared [`Mixer` class](https://github.com/google/oboe/blob/master/samples/shared/Mixer.h). [Example usage here. ](https://github.com/google/oboe/blob/master/samples/MegaDrone/src/main/cpp/Synth.h#L42)

> An app I am working on has the following use case...

It sounds like a karaoke app where the user listens to a song through headphones and sings into the microphone to create a voice recording. The voice recording is then mixed with the original song. 

If that's the use case then it's pretty simple. You need an output stream to play the song, and an input stream to record the microphone. Once the song is finished, iterate through the recording, mixing it with the original song, taking into account any [latency offset](https://github.com/google/oboe/blob/master/samples/hello-oboe/src/main/cpp/HelloOboeEngine.cpp#L42).
